packages/teams-js/src/internal/videoPerformanceStatistics.ts

```diff
@@ -0,0 +1,114 @@
+import { VideoFrameTick } from './videoFrameTick';
+
+export type VideoPerformanceStatisticsResult = {
+  effectId: string;
+  frameWidth: number;
+  frameHeight: number;
+  duration: number; // the duration in milliseconds that the data were collected
+  sampleCount: number; // the number of frames that were processed in the duration
+  distributionBins: Uint32Array; // an array that presents counts of frames that were finished in n milliseconds: distributionBins[frameProcessingDurationInMs]=frameCount. For example, distributionBins[10] = 5 means that 5 frames were processed in 10 milliseconds.
+};
+
+export class VideoPerformanceStatistics {
+  private static readonly initialSessionTimeoutInMs = 1000;
+  private static readonly maxSessionTimeoutInMs = 1000 * 30;
+
+  private currentSession: {
+    startedAtInMs: number;
+    timeoutInMs: number;
+    effectId: string;
+    frameWidth: number;
+    frameHeight: number;
+  };
+
+  private frameProcessingStartedAt: number;
+  private distributionBins: Uint32Array;
+  private sampleCount = 0;
+  private timeoutId: string;
+
+  public constructor(
+    distributionBinSize: number,
+    private report: (result: VideoPerformanceStatisticsResult) => void, // post event to the host
+  ) {
+    this.distributionBins = new Uint32Array(distributionBinSize);
+  }
+
+  /**
+   * Call this function before processing every frame
+   */
+  public processStarts(effectId: string, frameWidth: number, frameHeight: number): void {
+    VideoFrameTick.tick();
+    if (!this.suitableForThisSession(effectId, frameWidth, frameHeight)) {
+      this.reportAndResetSession(this.getStatistics(), effectId, frameWidth, frameHeight);
+    }
+    this.start();
+  }
+
+  public processEnds(): void {
+    // calculate duration of the process and record it
+    const durationInMillisecond = performance.now() - this.frameProcessingStartedAt;
+    const binIndex = Math.floor(Math.max(0, Math.min(this.distributionBins.length - 1, durationInMillisecond)));
+    this.distributionBins[binIndex] += 1;
+    this.sampleCount += 1;
+  }
+
+  private getStatistics(): VideoPerformanceStatisticsResult {
+    if (!this.currentSession) {
+      return null;
+    }
+    return {
+      effectId: this.currentSession.effectId,
+      frameHeight: this.currentSession.frameHeight,
+      frameWidth: this.currentSession.frameWidth,
+      duration: performance.now() - this.currentSession.startedAtInMs,
+      sampleCount: this.sampleCount,
+      distributionBins: this.distributionBins.slice(),
+    };
+  }
+
+  private start(): void {
+    this.frameProcessingStartedAt = performance.now();
```

Comment:
First time I've seen the use of `performance` global object. Is this supported everywhere?

Reply:
I think so, it should be supported by all the platforms we are using now:
![Uploading image.png…]()

https://developer.mozilla.org/en-US/docs/Web/API/Performance/now